### ISSUE TITLE
Datatype mismatch: 7 ERROR:  argument of AND must be type boolean, not type integer

### ISSUE THREAD

**Issue by blizzz** [nextcloud/circles#610]
With 22.0.0 beta 3 and Circles 22.0.0-alpha.6

Happens on every authenticated request, but might be user related seeing the backtrace.

```json
{
  "reqId": "NeUEGd56m92XQxEAQcyj",
  "level": 3,
  "time": "2021-06-02T20:45:02+00:00",
  "remoteAddr": "85.212.239.178",
  "user": "arthur",
  "app": "index",
  "method": "GET",
  "url": "/settings/apps/ocs/v2.php/apps/notifications/api/v2/notifications",
  "message": {
    "Exception": "Error",
    "Message": "Call to undefined method OCA\\Circles\\Db\\CoreQueryBuilder::validKey()",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/nextcloud/apps/circles/lib/Db/CoreQueryBuilder.php",
        "line": 586,
        "function": "generateAlias",
        "class": "OCA\\Circles\\Db\\CoreQueryBuilder",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/circles/lib/Db/CircleRequest.php",
        "line": 271,
        "function": "limitToDirectMembership",
        "class": "OCA\\Circles\\Db\\CoreQueryBuilder",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/circles/lib/Service/FederatedUserService.php",
        "line": 740,
        "function": "getSingleCircle",
        "class": "OCA\\Circles\\Db\\CircleRequest",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/circles/lib/Service/FederatedUserService.php",
        "line": 714,
        "function": "getSingleCircle",
        "class": "OCA\\Circles\\Service\\FederatedUserService",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/circles/lib/Service/FederatedUserService.php",
        "line": 345,
        "function": "fillSingleCircleId",
        "class": "OCA\\Circles\\Service\\FederatedUserService",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/circles/lib/ShareByCircleProvider.php",
        "line": 510,
        "function": "getLocalFederatedUser",
        "class": "OCA\\Circles\\Service\\FederatedUserService",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/lib/private/Share20/Manager.php",
        "line": 1353,
        "function": "getSharedWith",
        "class": "OCA\\Circles\\ShareByCircleProvider",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/files_sharing/lib/MountProvider.php",
        "line": 80,
        "function": "getSharedWith",
        "class": "OC\\Share20\\Manager",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/lib/private/Files/Config/MountProviderCollection.php",
        "line": 119,
        "function": "getMountsForUser",
        "class": "OCA\\Files_Sharing\\MountProvider",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/lib/private/Files/Filesystem.php",
        "line": 452,
        "function": "addMountForUser",
        "class": "OC\\Files\\Config\\MountProviderCollection",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/lib/private/Files/Filesystem.php",
        "line": 377,
        "function": "initMountPoints",
        "class": "OC\\Files\\Filesystem",
        "type": "::"
      },
      {
        "file": "/var/www/nextcloud/lib/private/legacy/OC_Util.php",
        "line": 325,
        "function": "init",
        "class": "OC\\Files\\Filesystem",
        "type": "::"
      },
      {
        "file": "/var/www/nextcloud/lib/base.php",
        "line": 1021,
        "function": "setupFS",
        "class": "OC_Util",
        "type": "::"
      },
      {
        "file": "/var/www/nextcloud/index.php",
        "line": 37,
        "function": "handleRequest",
        "class": "OC",
        "type": "::"
      }
    ],
    "File": "/var/www/nextcloud/apps/circles/lib/Db/CoreQueryBuilder.php",
    "Line": 1380,
    "CustomMessage": "--"
  },
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:88.0) Gecko/20100101 Firefox/88.0",
  "version": "22.0.0.6"
}
```

@daita

**Comment by blizzz** [comment]
seems to be some left over or not migrated from the TArrayTools? 

```
# grep  -r validKey
lib/Db/CoreQueryBuilder.php:            if (!$this->validKey($path, self::$SQL_PATH)
vendor/daita/my-small-php-tools/lib/Traits/TArrayTools.php:     public function validKey(string $k, array $arr): bool {
vendor/daita/my-small-php-tools/lib/Traits/TArrayTools.php:
```

**Comment by blizzz** [comment]
On first sight I also wonder why the trait's method is not being picked up

**Comment by blizzz** [comment]
This might have been an dependency issue with  files_lock that imports TArrayTools as well. I disabled and removed this, but now I have different errors, this time about database. We use postgres there.

```
{
  "reqId": "RC5eM8fp7uG4BFf4Xrrx",
  "level": 3,
  "time": "2021-06-03T09:00:26+00:00",
  "remoteAddr": "85.212.208.161",
  "user": "arthur",
  "app": "index",
  "method": "GET",
  "url": "/undefined/img/notifications.svg",
  "message": {
    "Exception": "Doctrine\\DBAL\\Exception\\DriverException",
    "Message": "An exception occurred while executing a query: SQLSTATE[42804]: Datatype mismatch: 7 ERROR:  argument of AND must be type boolean, not type integer\nLINE 1: ... = $3) AND (\"single_member\".\"level\" >= $4)) AND ((\"single\".\"...\n                                                             ^",
    "Code": 7,
    "Trace": [
      {
        "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Connection.php",
        "line": 1728,
        "function": "convert",
        "class": "Doctrine\\DBAL\\Driver\\API\\PostgreSQL\\ExceptionConverter",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Connection.php",
        "line": 1667,
        "function": "handleDriverException",
        "class": "Doctrine\\DBAL\\Connection",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Connection.php",
        "line": 1040,
        "function": "convertExceptionDuringQuery",
        "class": "Doctrine\\DBAL\\Connection",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/private/DB/Connection.php",
        "line": 232,
        "function": "executeQuery",
        "class": "Doctrine\\DBAL\\Connection",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php",
        "line": 210,
        "function": "executeQuery",
        "class": "OC\\DB\\Connection",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/private/DB/QueryBuilder/QueryBuilder.php",
        "line": 287,
        "function": "execute",
        "class": "Doctrine\\DBAL\\Query\\QueryBuilder",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/vendor/daita/my-small-php-tools/lib/Db/Nextcloud/nc22/NC22ExtendedQueryBuilder.php",
        "line": 1126,
        "function": "execute",
        "class": "OC\\DB\\QueryBuilder\\QueryBuilder",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/vendor/daita/my-small-php-tools/lib/Db/Nextcloud/nc22/NC22ExtendedQueryBuilder.php",
        "line": 1073,
        "function": "getRow",
        "class": "daita\\MySmallPhpTools\\Db\\Nextcloud\\nc22\\NC22ExtendedQueryBuilder",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Db/CircleRequestBuilder.php",
        "line": 108,
        "function": "asItem",
        "class": "daita\\MySmallPhpTools\\Db\\Nextcloud\\nc22\\NC22ExtendedQueryBuilder",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Db/CircleRequest.php",
        "line": 275,
        "function": "getItemFromRequest",
        "class": "OCA\\Circles\\Db\\CircleRequestBuilder",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
        "line": 740,
        "function": "getSingleCircle",
        "class": "OCA\\Circles\\Db\\CircleRequest",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
        "line": 714,
        "function": "getSingleCircle",
        "class": "OCA\\Circles\\Service\\FederatedUserService",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
        "line": 345,
        "function": "fillSingleCircleId",
        "class": "OCA\\Circles\\Service\\FederatedUserService",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/ShareByCircleProvider.php",
        "line": 510,
        "function": "getLocalFederatedUser",
        "class": "OCA\\Circles\\Service\\FederatedUserService",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/private/Share20/Manager.php",
        "line": 1353,
        "function": "getSharedWith",
        "class": "OCA\\Circles\\ShareByCircleProvider",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/apps/files_sharing/lib/MountProvider.php",
        "line": 80,
        "function": "getSharedWith",
        "class": "OC\\Share20\\Manager",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Config/MountProviderCollection.php",
        "line": 119,
        "function": "getMountsForUser",
        "class": "OCA\\Files_Sharing\\MountProvider",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Filesystem.php",
        "line": 452,
        "function": "addMountForUser",
        "class": "OC\\Files\\Config\\MountProviderCollection",
        "type": "->"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Filesystem.php",
        "line": 377,
        "function": "initMountPoints",
        "class": "OC\\Files\\Filesystem",
        "type": "::"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/private/legacy/OC_Util.php",
        "line": 325,
        "function": "init",
        "class": "OC\\Files\\Filesystem",
        "type": "::"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/lib/base.php",
        "line": 1021,
        "function": "setupFS",
        "class": "OC_Util",
        "type": "::"
      },
      {
        "file": "/var/www/tech-preview.nextcloud.com/index.php",
        "line": 37,
        "function": "handleRequest",
        "class": "OC",
        "type": "::"
      }
    ],
    "File": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Driver/API/PostgreSQL/ExceptionConverter.php",
    "Line": 83,
    "Previous": {
      "Exception": "Doctrine\\DBAL\\Driver\\PDO\\Exception",
      "Message": "SQLSTATE[42804]: Datatype mismatch: 7 ERROR:  argument of AND must be type boolean, not type integer\nLINE 1: ... = $3) AND (\"single_member\".\"level\" >= $4)) AND ((\"single\".\"...\n                                                             ^",
      "Code": 7,
      "Trace": [
        {
          "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php",
          "line": 84,
          "function": "new",
          "class": "Doctrine\\DBAL\\Driver\\PDO\\Exception",
          "type": "::"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Connection.php",
          "line": 1030,
          "function": "execute",
          "class": "Doctrine\\DBAL\\Driver\\PDO\\Statement",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/private/DB/Connection.php",
          "line": 232,
          "function": "executeQuery",
          "class": "Doctrine\\DBAL\\Connection",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php",
          "line": 210,
          "function": "executeQuery",
          "class": "OC\\DB\\Connection",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/private/DB/QueryBuilder/QueryBuilder.php",
          "line": 287,
          "function": "execute",
          "class": "Doctrine\\DBAL\\Query\\QueryBuilder",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/vendor/daita/my-small-php-tools/lib/Db/Nextcloud/nc22/NC22ExtendedQueryBuilder.php",
          "line": 1126,
          "function": "execute",
          "class": "OC\\DB\\QueryBuilder\\QueryBuilder",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/vendor/daita/my-small-php-tools/lib/Db/Nextcloud/nc22/NC22ExtendedQueryBuilder.php",
          "line": 1073,
          "function": "getRow",
          "class": "daita\\MySmallPhpTools\\Db\\Nextcloud\\nc22\\NC22ExtendedQueryBuilder",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Db/CircleRequestBuilder.php",
          "line": 108,
          "function": "asItem",
          "class": "daita\\MySmallPhpTools\\Db\\Nextcloud\\nc22\\NC22ExtendedQueryBuilder",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Db/CircleRequest.php",
          "line": 275,
          "function": "getItemFromRequest",
          "class": "OCA\\Circles\\Db\\CircleRequestBuilder",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
          "line": 740,
          "function": "getSingleCircle",
          "class": "OCA\\Circles\\Db\\CircleRequest",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
          "line": 714,
          "function": "getSingleCircle",
          "class": "OCA\\Circles\\Service\\FederatedUserService",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
          "line": 345,
          "function": "fillSingleCircleId",
          "class": "OCA\\Circles\\Service\\FederatedUserService",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/ShareByCircleProvider.php",
          "line": 510,
          "function": "getLocalFederatedUser",
          "class": "OCA\\Circles\\Service\\FederatedUserService",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/private/Share20/Manager.php",
          "line": 1353,
          "function": "getSharedWith",
          "class": "OCA\\Circles\\ShareByCircleProvider",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/apps/files_sharing/lib/MountProvider.php",
          "line": 80,
          "function": "getSharedWith",
          "class": "OC\\Share20\\Manager",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Config/MountProviderCollection.php",
          "line": 119,
          "function": "getMountsForUser",
          "class": "OCA\\Files_Sharing\\MountProvider",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Filesystem.php",
          "line": 452,
          "function": "addMountForUser",
          "class": "OC\\Files\\Config\\MountProviderCollection",
          "type": "->"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Filesystem.php",
          "line": 377,
          "function": "initMountPoints",
          "class": "OC\\Files\\Filesystem",
          "type": "::"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/private/legacy/OC_Util.php",
          "line": 325,
          "function": "init",
          "class": "OC\\Files\\Filesystem",
          "type": "::"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/lib/base.php",
          "line": 1021,
          "function": "setupFS",
          "class": "OC_Util",
          "type": "::"
        },
        {
          "file": "/var/www/tech-preview.nextcloud.com/index.php",
          "line": 37,
          "function": "handleRequest",
          "class": "OC",
          "type": "::"
        }
      ],
      "File": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Driver/PDO/Exception.php",
      "Line": 26,
      "Previous": {
        "Exception": "PDOException",
        "Message": "SQLSTATE[42804]: Datatype mismatch: 7 ERROR:  argument of AND must be type boolean, not type integer\nLINE 1: ... = $3) AND (\"single_member\".\"level\" >= $4)) AND ((\"single\".\"...\n                                                             ^",
        "Code": "42804",
        "Trace": [
          {
            "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php",
            "line": 82,
            "function": "execute",
            "class": "PDOStatement",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Connection.php",
            "line": 1030,
            "function": "execute",
            "class": "Doctrine\\DBAL\\Driver\\PDO\\Statement",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/private/DB/Connection.php",
            "line": 232,
            "function": "executeQuery",
            "class": "Doctrine\\DBAL\\Connection",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php",
            "line": 210,
            "function": "executeQuery",
            "class": "OC\\DB\\Connection",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/private/DB/QueryBuilder/QueryBuilder.php",
            "line": 287,
            "function": "execute",
            "class": "Doctrine\\DBAL\\Query\\QueryBuilder",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/vendor/daita/my-small-php-tools/lib/Db/Nextcloud/nc22/NC22ExtendedQueryBuilder.php",
            "line": 1126,
            "function": "execute",
            "class": "OC\\DB\\QueryBuilder\\QueryBuilder",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/vendor/daita/my-small-php-tools/lib/Db/Nextcloud/nc22/NC22ExtendedQueryBuilder.php",
            "line": 1073,
            "function": "getRow",
            "class": "daita\\MySmallPhpTools\\Db\\Nextcloud\\nc22\\NC22ExtendedQueryBuilder",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Db/CircleRequestBuilder.php",
            "line": 108,
            "function": "asItem",
            "class": "daita\\MySmallPhpTools\\Db\\Nextcloud\\nc22\\NC22ExtendedQueryBuilder",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Db/CircleRequest.php",
            "line": 275,
            "function": "getItemFromRequest",
            "class": "OCA\\Circles\\Db\\CircleRequestBuilder",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
            "line": 740,
            "function": "getSingleCircle",
            "class": "OCA\\Circles\\Db\\CircleRequest",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
            "line": 714,
            "function": "getSingleCircle",
            "class": "OCA\\Circles\\Service\\FederatedUserService",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/Service/FederatedUserService.php",
            "line": 345,
            "function": "fillSingleCircleId",
            "class": "OCA\\Circles\\Service\\FederatedUserService",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/circles/lib/ShareByCircleProvider.php",
            "line": 510,
            "function": "getLocalFederatedUser",
            "class": "OCA\\Circles\\Service\\FederatedUserService",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/private/Share20/Manager.php",
            "line": 1353,
            "function": "getSharedWith",
            "class": "OCA\\Circles\\ShareByCircleProvider",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/apps/files_sharing/lib/MountProvider.php",
            "line": 80,
            "function": "getSharedWith",
            "class": "OC\\Share20\\Manager",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Config/MountProviderCollection.php",
            "line": 119,
            "function": "getMountsForUser",
            "class": "OCA\\Files_Sharing\\MountProvider",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Filesystem.php",
            "line": 452,
            "function": "addMountForUser",
            "class": "OC\\Files\\Config\\MountProviderCollection",
            "type": "->"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/private/Files/Filesystem.php",
            "line": 377,
            "function": "initMountPoints",
            "class": "OC\\Files\\Filesystem",
            "type": "::"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/private/legacy/OC_Util.php",
            "line": 325,
            "function": "init",
            "class": "OC\\Files\\Filesystem",
            "type": "::"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/lib/base.php",
            "line": 1021,
            "function": "setupFS",
            "class": "OC_Util",
            "type": "::"
          },
          {
            "file": "/var/www/tech-preview.nextcloud.com/index.php",
            "line": 37,
            "function": "handleRequest",
            "class": "OC",
            "type": "::"
          }
        ],
        "File": "/var/www/tech-preview.nextcloud.com/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php",
        "Line": 82
      }
    },
    "CustomMessage": "--"
  },
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:88.0) Gecko/20100101 Firefox/88.0",
  "version": "22.0.0.6"
}
```

**Comment by tcitworld** [comment]
Closed by https://github.com/nextcloud/circles/commit/61f7c77339128a6e59c1e6af3ebedb248ebab9e8 ?

**Comment by blizzz** [comment]
> Closed by [61f7c77](https://github.com/nextcloud/circles/commit/61f7c77339128a6e59c1e6af3ebedb248ebab9e8) ?

aye
